services/airflow/src/dags/datenspende/add_vaccination_data/add_vaccination_data.py

```diff
@@ -0,0 +1,480 @@
+import pandas as pd
+import ramda as R
+import numpy as np
+from datetime import datetime
+from psycopg2.sql import SQL
+from src.lib.dag_helpers import (
+    execute_query_and_return_dataframe,
+    connect_to_db_and_upsert_pandas_dataframe_on_constraint,
+)
+from src.lib.test_helpers import set_env_variable_from_dag_config_if_present
+from postgres_helpers import create_db_context, teardown_db_context, DBContext
+
+
+def add_vaccination_data_to_homogenized_feature_table(**kwargs):
+    R.pipe(
+        set_env_variable_from_dag_config_if_present("TARGET_DB"),
+        load_and_transform_vaccination_data_for_each_user,
+        connect_to_db_and_upsert_pandas_dataframe_on_constraint(
+            "datenspende_derivatives", "homogenized_features", ["unique_answers"]
+        ),
+    )(kwargs)
+
+
+def load_and_transform_vaccination_data_for_each_user(*_):
+    db_context = create_db_context()
+    vaccination_data = load_vaccinations(db_context)
+    feature_data = load_feature_data(db_context)
+    teardown_db_context(db_context)
+
+    return R.pipe(
+        R.invoker(0, "itertuples"),
+        R.reduce(vaccination_data_reducer(vaccination_data), []),
+        pd.DataFrame,
+    )(feature_data)
+
+
+@R.curry
+def vaccination_data_reducer(vaccination_data, res, feature_data_row):
```

Comment:
I know it is tedeious, but I'd appreciate some docstrings if it is not immediately clear what a function does. At least one sentence of explanation and a definition of the input / output variables.